--- a/drivers/gpio/gpiolib.c
+++ b/drivers/gpio/gpiolib.c
@@ -47,7 +47,7 @@
  * Otherwise, minimize overhead in what may be bitbanging codepaths.
  */
 #ifdef	DEBUG
-#define	extra_checks	1
+#define	extra_checks	0
 #else
 #define	extra_checks	0
 #endif
@@ -2411,7 +2411,7 @@ int gpiod_get_raw_value(const struct gpi
 {
 	VALIDATE_DESC(desc);
 	/* Should be using gpio_get_value_cansleep() */
-	WARN_ON(desc->gdev->chip->can_sleep);
+	//WARN_ON(desc->gdev->chip->can_sleep);
 	return _gpiod_get_raw_value(desc);
 }
 EXPORT_SYMBOL_GPL(gpiod_get_raw_value);
